### DISCUSSION TITLE
integrated virus scan?

### DISCUSSION THREAD

**Discussion by harridu** [minio/minio#16350] [Q&A] [answered]
I'd love to replace nextcloud by minio as our upload and download server. Problem: Scanning for viruses and malware seems to require a notification mechanism to an external service doing the job. Error-prone. Painful. No sample implementation available. And it might collide with the integrated encryption, AFAICT.

Would it be possible to integrate an optional virus scan during upload?

Actually I thought I am not the first one asking for such a feature, but looking at the closed issues on github I haven't found similar requests before. Hopefully I wasn't too blind to see.

**Comment by harshavardhana** [comment, marked as answer]
@harridu scanning for anti viruses is an external job.

You first scan and then upload if it's not a virus. There is nothing MinIO needs to do here.

**Reply by manuschillerdev** [reply]
how would the flow be with pre-signed urls and client-uploads? 
Wouldn't it be easier to upload, scan and tag in this scenario?

**Reply by klauspost** [reply]
Correct. Or download through a [lambda function](https://github.com/minio/minio/tree/master/docs/lambda).

**Reply by manuschillerdev** [reply]
1. client: uploads via presigned-urls
2. minio: emits event on success
3. consumer: reacts to event and downloads and scans the file chunk by chunk (~1MB):
https://min.io/docs/minio/linux/developers/javascript/API.html#getPartialObject, https://www.npmjs.com/package/clamscan (`scanStream`, no page anchor available)
4. consumer flags file with the result when finished

would that be a possible flow? @klauspost 

thanks!

**Reply by klauspost** [reply]
@manuschillerdev Yes. That is our recommended solution for AV scans.

You should be able to make download policies dependent on AV scan tags, so files cannot be downloaded until the file has been tagged.

**Reply by manuschillerdev** [reply]
wonderful, thank you!

**Reply by pjbr** [reply]
Thanks!

**Comment by harridu** [comment]
The external scan would be some kind of recommendation or policy. Users are unreliable in this regard. Running an upload/download server I have to make sure that every uploaded document has been scanned, before it can be made available for download. Best way would be to refuse an infected file at upload time, instead of moving it to the destination folder in the bucket (IMHO).

I don't want to suggest a mandatory scan for all files in all buckets. Virus scan should be optional for each bucket (pretty similar to the shiny new Object Lifecycle Management feature). It would be an internal security feature.

**Reply by klauspost** [reply]
This is unlikely to become a core feature. It is however not an impossible piece of software to write.

You can set up notifications for uploads: https://min.io/docs/minio/kubernetes/upstream/administration/monitoring.html

This can be picked up by a scanning service, which downloads, scans and applies a [tag](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectTagging.html)  with the result. This means that un-tagged objects are un-scanned, but if you don't want to move objects from a staging area, then that is your option.

If the service keeps track of the objects, it can rescan at desired intervals.
